OpenSim/Simulation/StatesTrajectory.h

```diff
@@ -221,6 +230,164 @@ class OSIMSIMULATION_API StatesTrajectory {
+     * the state at time 1.3. You could remedy this issue by setting the
+     * tolerance to 1e-3 (the precision of the console output). So, this
+     * function actually gives the nearest state before `time + tolerance`.
+     *
+     * For a trajectory with times 0.2, 0.5, 0.7, this code will get the state
+     * at time 0.5:
+     * @code
+     * const auto& state = states[states.findIndexNearestBefore(0.6)];
+     * @endcode
+     *
+     * @throws TimeOutOfRange If there are no states before the given time.
+     * @throws Exception If the trajectory is empty.
+     */
+    size_t findIndexNearestBefore(const double& time,
+                                  const double& tolerance = 0) const;
+    /** Get the index of the nearest state after the given time
+     * (possibly *at* the given time).
+     *
+     * See findIndexNearestBefore() for an explanation of the tolerance; this
+     * function gives the nearest state after `time - tolerance`.
+     *
+     * For a trajectory with times 0.2, 0.5, 0.7, this code will get the state
+     * at time 0.5:
+     * @code
+     * const auto& state = states[states.findIndexNearestAfter(0.4)];
+     * @endcode
+     *
+     * @throws TimeOutOfRange If there are no states after the given time.
+     * @throws Exception If the trajectory is empty.
+     */
+    size_t findIndexNearestAfter(const double& time,
+                                 const double& tolerance = 0) const;
+
+    /** Get the index of the state closest to the given time within the
+     * interval [time - tolerance, time + tolerance].
+     * If there are multiple states exactly at the given time, the
+     * index is that of the first of these states.
+     *
+     * This tolerance is *different* from the one used in
+     * findIndexNearestBefore(), but should still be set to the precision of
+     * the desired time where you observed it (perhaps in console output or
+     * storage files).
+     *
+     * For a trajectory with times 0.2, 0.501, 0.7, this code gives the index
+     * for the state at time 0.5:
+     * @code
+     * states.findIndexAt(0.5, 0.01);
+     * @endcode
+     *
+     * @throws NoStateAtTime If there is no state at the given time.
+     * @throws Exception If the trajectory is empty.
+     */
+    size_t findIndexAt(const double& time, const double& tolerance = 0) const;
+    /** Iterate over the states in a given time interval (inclusive).
+     *
+     * For a trajectory with times 0.2, 0.5, 0.7, 1.5, 1.8, this code will
+     * iterate through the states at times 0.5, 0.7, 1.5:
+     * @code
+     * for (const auto& state : states.findBetween(0.5, 1.5)) {
+     *     state.getTime();
+     * }
+     * @endcode
+     *
+     * The startTime must be less than or equal to the endTime, otherwise an
+     * exception is thrown.
+     *
+     * See findIndexNearestBefore() for an explanation of the tolerance.
+     */
+    IteratorRange findBetween(const double& startTime, const double& endTime,
```

Comment:
Yes, but based on our discussion yesterday, I'll change this so that if there are multiple states between `startTime - tolerance` and `startTime`, I'll use the one closest to `startTime` within that interval. And likewise for the last state.